src/checkers/inference/solver/backend/z3smt/Z3SmtSolver.java

```diff
@@ -0,0 +1,429 @@
+package checkers.inference.solver.backend.z3smt;
+
+import checkers.inference.model.ArithmeticConstraint;
+import checkers.inference.model.ArithmeticConstraint.ArithmeticOperationKind;
+import checkers.inference.model.Constraint;
+import checkers.inference.model.Slot;
+import checkers.inference.model.serialization.ToStringSerializer;
+import checkers.inference.solver.backend.Solver;
+import checkers.inference.solver.backend.z3smt.encoder.Z3SmtSoftConstraintEncoder;
+import checkers.inference.solver.frontend.Lattice;
+import checkers.inference.solver.util.ExternalSolverUtils;
+import checkers.inference.solver.util.FileUtils;
+import checkers.inference.solver.util.SolverArg;
+import checkers.inference.solver.util.SolverEnvironment;
+import checkers.inference.solver.util.Statistics;
+import com.microsoft.z3.BoolExpr;
+import com.microsoft.z3.Context;
+import com.microsoft.z3.Expr;
+
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.logging.Logger;
+import javax.lang.model.element.AnnotationMirror;
+import org.checkerframework.javacutil.BugInCF;
+
+public class Z3SmtSolver<SlotEncodingT, SlotSolutionT>
+        extends Solver<Z3SmtFormatTranslator<SlotEncodingT, SlotSolutionT>> {
+
+    public enum Z3SolverEngineArg implements SolverArg {
+        /** option to use optimizing mode or not */
+        optimizingMode
+    }
+
+    private static final Logger logger = Logger.getLogger(Z3SmtSolver.class.getName());
+
+    protected final Context ctx;
+    protected com.microsoft.z3.Optimize solver;
+    protected StringBuilder smtFileContents;
```

Comment:
I don't think this needs to be a field. One method sets it and a few helper methods can take it as argument. That would make the flow much clearer.

Reply:
I also think it makes the flow clearer to pass the smt content in and out of every step in the solving process.

But now this field is written in `void encodeAllConstraints()`, which is a common interface implemented in other solvers. Some of the solvers are not string-based and we cannot just simply add string parameter and return.